Support hexagon tiles in `GridMap`

octetdev2 opened the pull request:
This adds support for hexagonal grids in the `GridMap` node with the same offset and layout semantics as `TileMap`.

Closes https://github.com/godotengine/godot-proposals/issues/4337

<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

caspervonb committed:
Support hexagonal cells in `GridMap`

This adds support for hexagonal grids in the `GridMap` node with the
same offset and layout semantics as `TileMap`.

fire-forge reviewed (commented):
I tested it and it works really well. Just a few things need to be changed/fixed:
* The editor grid drawing should be changed to respect `cell_shape` and `cell_offset_axis` settings like the TileMap editor grid does.
* `cell_offset_axis` only works in hexagon mode, so you can't use offset squares like you can in TileMaps.
* The placement position in the editor is off by about half of a cell. This is more apparent when you go into the Top Orthogonal camera mode when drawing on the GridMap. Disabling `cell_center_x/y/z` makes it off by half of a cell in the opposite direction.

octetdev2 commented:
> The editor grid drawing should be changed to respect cell_shape and cell_offset_axis settings like the TileMap editor grid does.

Was going to split this out into a follow up pull request, but squashed it into here.

> cell_offset_axis only works in hexagon mode, so you can't use offset squares like you can in TileMaps.

Yes, there is no half offset square shape yet.
But the inspector should disable the options when not applicable.

> The placement position in the editor is off by about half of a cell. This is more apparent when you go into the Top Orthogonal camera mode when drawing on the GridMap. Disabling cell_center_x/y/z makes it off by half of a cell in the opposite direction.

I can't reproduce this? The transform for cursor instance and rendering instance should be identical.

fire-forge commented:
> I can't reproduce this? The transform for cursor instance and rendering instance should be identical.

I meant that the mouse cursor position vs. which cell it decides you're hovering over is about half a cell off, not the cursor and rendering instance being off. I'll try to get a video of it, and I'll check if it happened before this PR too.

octetdev2 commented:
>Nevermind, glad to hear it's fixed.

Actually, I lost the grid offset changes when I rebased and squashed about an hour ago... never the less, easy fix.

octetdev2 commented on modules/gridmap/grid_map.cpp:
```diff
@@ -409,18 +436,231 @@ int GridMap::get_cell_item_orientation(const Vector3i &p_position) const {
+
+	// Add the offset
+	Vector3 offset = Vector3(
+			cell_size.x * 0.5 * int(center_x),
```
```suggestion
			0.5 * int(center_x),
```

octetdev2 commented on modules/gridmap/grid_map.cpp:
```diff
@@ -409,18 +436,231 @@ int GridMap::get_cell_item_orientation(const Vector3i &p_position) const {
+	// Add the offset
+	Vector3 offset = Vector3(
+			cell_size.x * 0.5 * int(center_x),
+			cell_size.y * 0.5 * int(center_y),
```
```suggestion
			0.5 * int(center_y),
```

octetdev2 commented on modules/gridmap/grid_map.cpp:
```diff
@@ -409,18 +436,231 @@ int GridMap::get_cell_item_orientation(const Vector3i &p_position) const {
+	Vector3 offset = Vector3(
+			cell_size.x * 0.5 * int(center_x),
+			cell_size.y * 0.5 * int(center_y),
+			cell_size.z * 0.5 * int(center_z));
```
```suggestion
			0.5 * int(center_z));
```

octetdev2 reviewed (commented):
Oops, multiplying this twice would be the cause of that extra offset (which was only observable depending on cell size).

octetdev2 commented:
- Figured out the property validator and disabled the properties for square cells.
- Also squashed the off by one error when picking (off by one extra pass of multiplication 😆).

Wasabi-Cheetah commented:
I'm just going to post this here for anyone else looking for a hexagon grid system for a sphere as well. https://stackoverflow.com/questions/46777626/mathematically-producing-sphere-shaped-hexagonal-grid

soft-circles commented:
I would love to see this feature merged, any chance this will make it in during the Godot 4 beta?

Calinou commented:
> I would love to see this feature merged, any chance this will make it in during the Godot 4 beta?

[Godot 4.0 is in feature freeze, so any features of a significant size are postponed to future 4.x releases.](https://godotengine.org/article/release-management-4-0-and-beyond)

gulagkulak commented:
@octetdev2 seems to have disappared from the internet (I hope he's ok) so, uh, can someone please take over development of this feature?

AThousandShips commented:
Salvaged in: #85890
